File: samples/sample.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

// Class
class FooBar {
public:
    std::string foo;
    int bar;

    FooBar(std::string foo, int bar) : foo(foo), bar(bar) {}

    void print() {
        std::cout << "Foo: " << foo << ", Bar: " << bar << std::endl;
    }
};

// Function
void printVector(const std::vector<int>& vec) {
    for (int i : vec) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

int main() {
    // Using the class
    FooBar foobar("Lorem Ipsum", 123);
    foobar.print();

    // Using a vector
    std::vector<int> baz = {1, 2, 3, 4, 5};
    printVector(baz);

    // Using a map
    std::map<std::string, std::string> m;
    m["foo"] = "bar";
    m["baz"] = "qux";

    // Using control structures
    for (const auto& pair : m) {
        std::cout << "Key: " << pair.first << ", Value: " << pair.second << std::endl;
    }

    if (m.find("foo") != m.end()) {
        std::cout << "Found foo: " << m["foo"] << std::endl;
    } else {
        std::cout << "Did not find foo" << std::endl;
    }

    return 0;
}

File: samples/sample.go
package main

import (
    "fmt"
    "strings"
)

// Struct
type FooBar struct {
    Foo string
    Bar int
}

// Function
func NewFooBar(foo string, bar int) *FooBar {
    return &FooBar{Foo: foo, Bar: bar}
}

// Method
func (fb *FooBar) Print() {
    fmt.Printf("Foo: %s, Bar: %d\n", fb.Foo, fb.Bar)
}

// Interface
type Printer interface {
    Print()
}

// Function using interface
func PrintTwice(p Printer) {
    p.Print()
    p.Print()
}

func main() {
    // Using the struct and methods
    foobar := NewFooBar("Lorem Ipsum", 123)
    foobar.Print()

    // Using the interface
    PrintTwice(foobar)

    // Using the strings package
    fmt.Println(strings.Contains(foobar.Foo, "Lorem"))

    // Using a map
    m := make(map[string]string)
    m["foo"] = "bar"
    m["baz"] = "qux"
    fmt.Println(m)

    // Using control structures
    for key, value := range m {
        fmt.Printf("Key: %s, Value: %s\n", key, value)
    }

    if val, ok := m["foo"]; ok {
        fmt.Println("Found foo:", val)
    } else {
        fmt.Println("Did not find foo")
    }
}
